ls1intum/Artemis
# `Development`: Move programming exercise repository actions into separate service

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
The `ProgrammingExerciseService` class was very big.

### Description
This PR moves the repository actions from the `ProgrammingExerciseService` into a new separate `ProgrammingExerciseRepositoryService` to keep the classes smaller.
Additionally, some long methods were refactored and split up into smaller units.

The old methods using `String`s for path representation in the `ResourceLoaderService` have been deprecated in favour of representing paths with the intended `Path` objects.
To make reviewing this PR easier I did only mark them as `@Deprecated` for now. I already have planned on fully removing them in a follow-up PR. I’ll open a stacked PR for that to reduce the time between merging the PRs in which the deprecation warnings are shown by the compiler.

### Steps for Testing

Prerequisites:
- 1 Instructor

Things to test:

- Create programming exercises for the various programming languages and check that the repository is still set up as expected, i.e. the builds for the template/solution repositories run through, and tests pass/fail accordingly.
- Java and Swift exercises should be checked thoroughly, since they allow for various combinations of build tools and features.
  - When enabling static code analysis on exercise creating, and adding an issue to e.g. the solution, a feedback should show up in Artemis.
  - The respective build-tools files should be part of the repositories (e.g. `pom.xml` for Maven, `build.gradle` for Gradle).
  - The builds should work when enabling sequential test runs (to be tested for Kotlin as well).
  - The builds should work when enabling testwise coverage (only Java).


#### Exam Mode Testing
No special testing needed, if the repository setup still works for regular programming exercises, then it works for exams, too, since the mechanism is identical.

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage



| Class/File | Branch | Line |
|------------|-------:|-----:|
| ProgrammingExerciseService.java | 76% | 93% |
| ProgrammingExerciseRepositoryService.java | 82% | 92% |
